go-controller/pkg/network-attach-def-controller/network_manager.go

```diff
@@ -158,8 +162,26 @@ func (nm *networkManagerImpl) syncAll() error {
 	// as we sync upon start, consider networks that have not been ensured as
 	// stale and clean them up
 	validNetworks := make([]util.BasicNetInfo, 0, len(nm.networks))
-	for _, network := range nm.networks {
+	networkNames := make([]string, 0, len(nm.networks))
+	for name, network := range nm.networks {
 		validNetworks = append(validNetworks, network)
+		networkNames = append(networkNames, name)
+	}
+	if err := nm.ncm.CleanupDeletedNetworks(validNetworks...); err != nil {
+		return err
 	}
-	return nm.ncm.CleanupDeletedNetworks(validNetworks...)
+
+	// sync all known networks. There is no informer for networks. Keys are added by NAD controller.
+	// Therefore, there is no serialization for initial adds, since addHandler is not called and therefore
+	// networkManager has a non-blocking start up. Leverage syncAll to block NetworkManager start up to
+	// simulate addHandler behavior, waiting for all networks to be added.
```

Comment:
I don't think this comment really explains why this is needed. I think the key here is that existing networks need to be started synchronously, maybe this is what you mean with serialization? (adds are serialized because we only have a single thread in the controller). 

Perhaps this other section of your explanation is the better perspective
```
For certain controllers that integrate with UDN, they contain caches, which depend on being initialized by networks having been created at start up. If those controllers were to start without a fully initialized cache, they may end up temporarily deleting configuration that could cause disruption.
```

You talk about IP rule manager. I guess that is shared among networks and if it starts up without being aware of them it removes configuration it thinks as stale. I guess this works but all these things that will be shared by multiple networks would really need to be managed at a higher level, like maybe by network manager itself?

Anyway, I would probably edit the this comment to be something similar to
```
Certain downstream controllers that handle configuration for multiple networks depend on being aware of all the existing networks on initialization. To achieve that, we need to start existing networks synchronously. Otherwise these controllers might incorrectly incorrectly asses certain configuration as stale"   
```

Reply:
imo with almost all controllers today we think about them as being fully initialized and handled all the adds before we move onto starting the next controller. The order matters for us with node handler, pod handler, etc. I think we should keep that consistent even with our network controller, unless there is some good justification. Otherwise I worry we will just confuse ourselves even more.